dvc/repo/experiments/queue/celery.py

```diff
@@ -359,7 +365,8 @@ def logs(
         ) as fobj:
             ui.write(fobj.read())
 
-    def worker_status(self) -> Dict:
+    @cached_property
+    def worker_status(self) -> Dict[str, List]:
```

Comment:
This should not be a `cached_property`.

If it's a `cached_property`, it means we will only get the worker state once, and we will continue to only return that fixed state for every subsequent call (meaning if tasks finish are more tasks are added to the worker, we will never return them).

Reply:
In places that we need to use the worker status, we should call the method once and assign it to a local variable, and then read/iterate from that variable as needed, like we were doing in the original line 148:

```
active_worker: Dict = self.worker_status()
```

This gives us the equivalent of the "cached" behavior within the function using that status at a given point in time, but means that we also properly handle when other tasks are added/removed in the future

Reply:
I'm afraid the `worker_status` takes about 1S to get a result. And in `exp_show` we called twice, it would add more than 2 seconds if we do not cache its result.

Reply:
I would not worry about that performance impact right now, if celery is slow, then we can live with that for now.

We can consider caching in the future to improve performance, but it would have to be a time based implementation (where we actually expire the cached result after some amount of time). Using `cached_property` is not the right way to handle this though.

Reply:
It may also be possible to adjust the `exp show` workflow for handling running tasks. If we can get rid of the second call to `worker_status`, that would also address the performance concern.

Reply:
Let's focus on stability first and performance later as needed.

Reply:
For the working status, I think it's OK if we use a universal result throughout one operation. And maybe it would make less confusion if there is a status change during the operation. For example, at first, we mark an experiment as queued in `_get_running_exps` and then as running when we collect running experiments in `exp show`. 

I removed all cache, and then the experience in the vs-code extension would be terrible. If you tried it on the extension you can feel the latency in all operations. 

![image](https://user-images.githubusercontent.com/6745454/202137165-a3af7581-c6ca-4a80-8551-cb24c6bfb54d.png)

If we can't add a cache, I would prefer not to check the running status in `exp show` and instead make users can kill a false shown running experiment.

Reply:
> I would prefer not to check the running status in exp show and instead make users can kill a false shown running experiment.

@karajan1001 This is acceptable as well. The original issues were about not being able to kill experiments that are in a pending/init celery state (whether due to the experiment still being initialized, or because a worker crash made the task stuck in that state). All we need to fix the issue is to do the full celery worker checks in `queue kill` (and remove the bad/stuck celery task messages if necessary).

We do not need to be doing the slower celery checks in `exp show`.

Reply:
> We do not need to be doing the slower celery checks in `exp show`, so you can keep the existing behavior for `exp show`/`queue status`

What is the existing behavior? How are statuses reported in `exp show`/`queue status` if not checking the celery worker state?

Is the idea here to sometimes report outdated status in `exp show`/`queue status` but ensure that `queue remove/kill` never fails even if someone tries to kill a "running" experiment that's actually being initialized or already completed? I agree that seems like a fine solution for now if it addresses the `queue remove/kill` failures and keeps `exp show`/`queue status` performance reasonable.

Reply:
> What is the existing behavior? How are statuses reported in `exp show`/`queue status` if not checking the celery worker state?

We only use the status stored in `msg` and do not examine if the process really exists. 

> Is the idea here to sometimes report outdated status in exp show/queue status but ensure that queue remove/kill never fails even if someone tries to kill a "running" experiment that's actually being initialized or already completed?

The original PR is to make the running experiment double-checked by process, and it will ensure that all the `running` experiment in `exp show/queue status` is actually active and can be killed. The original problem is that we might try to kill inactive tasks, which is impossible.

Reply:
Even if there is a double check, there could always be a race condition, right? Can we start with a way to clean up the inactive task if someone tries to kill it, which I think is what @pmrowla suggested above?

Reply:
@dberenbaum I had a call with @karajan1001 yesterday and we agreed to go with that. So if there's an invalid/inactive task we will handle cleaning it up on kill (instead of generating the errors from the original issue).